staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/admission.go

```diff
@@ -54,28 +61,87 @@ func Register(plugins *admission.Plugins) {
 
 type celAdmissionPlugin struct {
 	evaluator CELPolicyEvaluator
+	enabled   bool
+
+	// Injected Dependencies
+	informerFactory informers.SharedInformerFactory
+	client          kubernetes.Interface
+	restMapper      meta.RESTMapper
+	dynamicClient   dynamic.Interface
+	stopCh          <-chan struct{}
 }
 
-var _ WantsCELPolicyEvaluator = &celAdmissionPlugin{}
+var _ initializer.WantsExternalKubeInformerFactory = &celAdmissionPlugin{}
+var _ initializer.WantsExternalKubeClientSet = &celAdmissionPlugin{}
+var _ WantsRESTMapper = &celAdmissionPlugin{}
+var _ initializer.WantsDynamicClient = &celAdmissionPlugin{}
+var _ initializer.WantsDrainedNotification = &celAdmissionPlugin{}
+
+var _ admission.InitializationValidator = &celAdmissionPlugin{}
 var _ admission.ValidationInterface = &celAdmissionPlugin{}
 
-func NewPlugin() (*celAdmissionPlugin, error) {
+func NewPlugin() (admission.Interface, error) {
 	result := &celAdmissionPlugin{}
 	return result, nil
 }
 
-func (c *celAdmissionPlugin) SetCELPolicyEvaluator(evaluator CELPolicyEvaluator) {
-	c.evaluator = evaluator
+func (c *celAdmissionPlugin) SetExternalKubeInformerFactory(f informers.SharedInformerFactory) {
+	c.informerFactory = f
+}
+
+func (c *celAdmissionPlugin) SetExternalKubeClientSet(client kubernetes.Interface) {
+	c.client = client
+}
+
+func (c *celAdmissionPlugin) SetRESTMapper(mapper meta.RESTMapper) {
+	c.restMapper = mapper
+}
+
+func (c *celAdmissionPlugin) SetDynamicClient(client dynamic.Interface) {
+	c.dynamicClient = client
 }
 
-// Once clientset and informer factory are provided, creates and starts the
-// admission controller
+func (c *celAdmissionPlugin) SetDrainedNotification(stopCh <-chan struct{}) {
+	c.stopCh = stopCh
+}
+
+// ValidateInitialization - once clientset and informer factory are provided, creates and starts the admission controller
 func (c *celAdmissionPlugin) ValidateInitialization() error {
-	if c.evaluator != nil {
+	if !c.enabled {
 		return nil
 	}
+	if c.informerFactory == nil {
+		return errors.New("missing informer factory")
+	}
+	if c.client == nil {
+		return errors.New("missing kubernetes client")
+	}
+	if c.restMapper == nil {
+		return errors.New("missing rest mapper")
+	}
+	if c.dynamicClient == nil {
+		return errors.New("missing dynamic client")
+	}
+	if c.stopCh == nil {
+		return errors.New("missing stop channel")
+	}
+	c.evaluator = NewAdmissionController(c.informerFactory, c.client, c.restMapper, c.dynamicClient)
```

Comment:
Running this in my debugger showed this line got hit 3 times. This is because there is a plugin instance created for each `genericapiserver` in the server chain. This was the motivation behind the original integration strategy.

@liggitt do you have suggestions to have plugins cleanly share the `CELPolicyEvaluator` instance?

Reply:
I suppose we could have the initializer inject a shared instance to each plugin and keep the current plugin DI. But that brings complications around who is responsible for starting the controller, and whose injected dependencies get to stay on the object?

Reply:
huh... that's exciting... do we have the same issue with the webhook admission plugins? I guess the most expensive thing those do is make clients and compile selectors

Reply:
it actually looks like this is needed... see https://github.com/kubernetes/kubernetes/pull/61404 for when this was added

each internal server knows about conversions for its own types... until @jpbetz's fix, we were doing conversions to send objects to cel, but now that we are, I wonder if we would have hit issues trying to use CEL validation on APIService or CustomResourceDefinition objects when the conversions were done

this reminds me, we should add a test to the list of graduation requirements which registers a validation policy for everything and iterates through all groups/versions/resources/subresources and ensures they get intercepted and work properly with a CEL validation policy... that test turned up lots of weird edges for admission webhooks

Reply:
> If the thing used to convert is passed to admission in ObjectInterfaces, we might be able to use a shared compiled CEL controller…

Maybe I'm not following, but IIRC the converter passed in via `ObjectInterfaces` should only ever be used against objects retrieved from the request's `admission.Attributes` passed as its sibling to the plugin's `Validate` and it should not outlive the life of that function.

Reply:
The beta graduation criteria is tracked here: https://github.com/kubernetes/enhancements/pull/3644

Reply:
it looks like the reason for adding the duplicate admission constructions is no longer required - opened https://github.com/kubernetes/kubernetes/pull/113634 to drop them